src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs

```diff
@@ -133,25 +137,35 @@ async void InitializeAsync()
 
             //Due that the Web Browser(WebView2 - Chromium) security CORS is blocking the load of resources like images then we need to create a virtual folder in which the image are located.
             this.documentationBrowser.CoreWebView2.SetVirtualHostNameToFolderMapping(VIRTUAL_FOLDER_MAPPING, FallbackDirectoryName, CoreWebView2HostResourceAccessKind.DenyCors);
+            this.documentationBrowser.CoreWebView2.WebMessageReceived += CoreWebView2OnWebMessageReceived;
 
             string htmlContent = this.viewModel.GetContent();
 
             Dispatcher.BeginInvoke(new Action(() =>
             {
                 this.documentationBrowser.NavigateToString(htmlContent);
             }));
-
-            this.documentationBrowser.CoreWebView2.WebMessageReceived += CoreWebView2OnWebMessageReceived;
         }
 
+        // TODO: This event fires twice causing issues down the line
         private void CoreWebView2OnWebMessageReceived(object sender, CoreWebView2WebMessageReceivedEventArgs e)
         {
             var message = e.TryGetWebMessageAsString();
             if (string.Equals(message, "insert"))
             {
                 // Insert the graph inside the current worskspace
+                if (MessageIsHandled) return;
                 this.viewModel.InsertGraph();
             }
+
+            MessageIsHandled = true;
+            HandleInsert();
+        }
+
+        private async void HandleInsert()
+        {
+            await Task.Delay(MESSAGE_HANDLE_DELAY);
```

Comment:
I absolutely understand, and I was looking for an alternative way to solve this, but I cannot find a way around it. What I am observing is that [WebView2.WebMessageReceived event](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.webmessagereceived?view=webview2-dotnet-1.0.1418.22) is firing twice. In my case, this triggers the second insertion of a graph inside Dynamo, which is definitely not what we want.  Since I am raising a Warning and not allowing this to happen, the result is a faulty prompt being raised. The only way I managed to suppress this behavior was to introduce a 'timer' and only allow a single run every 2 seconds. The Task.Delay is inside an async method, so it's not halting the main thread, I am just timing out a boolean. I have been puzzling over this for quite a while, I don't know how else to solve it @mjkkirschner  if you have stumbled on this issue, or if you could suggest a more elegant way of handling this, I would do it!

Reply:
@dnenov can you give me more info on what message is triggering the handler? 

Maybe you can tell me a bit more about the design here - when a user hits insert what happens? 

I would assume you could skip this handler and use the scriptingObject directly to marshal your call to the c# side - I guess my question is - do you have to use this event or is there an alternative?

Reply:
Sure thing. To my understanding, this is what's happening:
- the DocumentationBrowser is an HTML page displayed through the WebView2 control inside the WPF Window
- user engagement happens on the HTML side. Clicking the 'Insert' button triggers a window.chrome.webview.postMessage(message) where the message is an arbitrary string
- we capture that message by subscribing to the 'WebView2.WebMessageReceived'. We use the message string just to guide us to the desired functionality, the rest of it is on the C# side
- when testing, the **postMessage** is only triggered once on the HTML side. However, the **WebMessageReceived** is triggered twice. I cannot find any documentation confirming this redundancy, but it is what I am observing.

What would be an alternative design pattern between the hosted HTML page and the WPF? At the moment, this seems to be resolving the issue, but I will change it to a more robust pattern if possible.

Reply:
you can also use `hostObjects` to call a method on a c# object.
https://github.com/DynamoDS/Dynamo/blob/master/src/LibraryViewExtensionWebView2/web/library/library.html#L157
https://github.com/DynamoDS/Dynamo/search?q=bridgeTwoWay

Though I do not believe we do not see the same double execution behavior with the postMessage command in librareieJS which is also an html/js application using webView2 so I would debug the startup code to make sure that you do not accidentally subscribe the same webView2 instance to this event twice - ie do you see two event handlers in the list of subscribers?

you can use quickwatch with `GetInvocationList()` on the event to check this.

Reply:
Hi @mjkkirschner - great news, @reddyashish and I finally solved this puzzle! What was happening was that the CoreWebView2 was being initialized every time a new Node help menu was opened. So by the second Node help, we would have subscribed 2, and then 3 times, and so on. 
Thank you for chasing me on this one, I think we solved a bigger problem by not allowing me to 'patch' things up. I will mark this one as resolved.

Reply:
awesome, great work guys.